File: lib/features/clan/models/clan_war_stats_filter.dart
class ClanWarStatsFilter {
  final DateTime? startDate;
  final DateTime? endDate;
  final int? ownTownHall;
  final int? enemyTownHall;
  final List<int>? ownTownHalls;
  final List<int>? enemyTownHalls;
  final bool sameTownHall;
  final String warType;
  final List<String>? warTypes;
  final bool? freshAttacksOnly;
  final int? minStars;
  final int? maxStars;
  final List<int>? allowedStars;
  final double? minDestruction;
  final double? maxDestruction;
  final int? minMapPosition;
  final int? maxMapPosition;
  final int limit;

  const ClanWarStatsFilter({
    this.startDate,
    this.endDate,
    this.ownTownHall,
    this.enemyTownHall,
    this.ownTownHalls,
    this.enemyTownHalls,
    this.sameTownHall = false,
    this.warType = "all",
    this.warTypes,
    this.freshAttacksOnly,
    this.minStars,
    this.maxStars,
    this.allowedStars,
    this.minDestruction,
    this.maxDestruction,
    this.minMapPosition,
    this.maxMapPosition,
    this.limit = 50,
  });

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {
      'limit': limit,
      'same_th': sameTownHall,
    };
    
    if (warTypes != null && warTypes!.isNotEmpty) {
      data['type'] = warTypes;
    } else {
      data['type'] = warType;
    }

    if (startDate != null) {
      data['timestamp_start'] = startDate!.millisecondsSinceEpoch ~/ 1000;
    }
    if (endDate != null) {
      data['timestamp_end'] = endDate!.millisecondsSinceEpoch ~/ 1000;
    }
    if (ownTownHalls != null && ownTownHalls!.isNotEmpty) {
      data['own_th'] = ownTownHalls;
    } else if (ownTownHall != null) {
      data['own_th'] = ownTownHall;
    }
    if (enemyTownHalls != null && enemyTownHalls!.isNotEmpty) {
      data['enemy_th'] = enemyTownHalls;
    } else if (enemyTownHall != null) {
      data['enemy_th'] = enemyTownHall;
    }
    if (freshAttacksOnly != null) {
      data['fresh_only'] = freshAttacksOnly;
    }
    if (allowedStars != null && allowedStars!.isNotEmpty) {
      data['stars'] = allowedStars;
    } else {
      if (minStars != null) {
        data['min_stars'] = minStars;
      }
      if (maxStars != null) {
        data['max_stars'] = maxStars;
      }
    }
    if (minDestruction != null) {
      data['min_destruction'] = minDestruction;
    }
    if (maxDestruction != null) {
      data['max_destruction'] = maxDestruction;
    }
    if (minMapPosition != null) {
      data['map_position_min'] = minMapPosition;
    }
    if (maxMapPosition != null) {
      data['map_position_max'] = maxMapPosition;
    }

    return data;
  }

  ClanWarStatsFilter copyWith({
    DateTime? startDate,
    DateTime? endDate,
    int? ownTownHall,
    int? enemyTownHall,
    List<int>? ownTownHalls,
    List<int>? enemyTownHalls,
    bool? sameTownHall,
    String? warType,
    List<String>? warTypes,
    bool? freshAttacksOnly,
    int? minStars,
    int? maxStars,
    List<int>? allowedStars,
    double? minDestruction,
    double? maxDestruction,
    int? minMapPosition,
    int? maxMapPosition,
    int? limit,
  }) {
    return ClanWarStatsFilter(
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      ownTownHall: ownTownHall ?? this.ownTownHall,
      enemyTownHall: enemyTownHall ?? this.enemyTownHall,
      ownTownHalls: ownTownHalls ?? this.ownTownHalls,
      enemyTownHalls: enemyTownHalls ?? this.enemyTownHalls,
      sameTownHall: sameTownHall ?? this.sameTownHall,
      warType: warType ?? this.warType,
      warTypes: warTypes ?? this.warTypes,
      freshAttacksOnly: freshAttacksOnly ?? this.freshAttacksOnly,
      minStars: minStars ?? this.minStars,
      maxStars: maxStars ?? this.maxStars,
      allowedStars: allowedStars ?? this.allowedStars,
      minDestruction: minDestruction ?? this.minDestruction,
      maxDestruction: maxDestruction ?? this.maxDestruction,
      minMapPosition: minMapPosition ?? this.minMapPosition,
      maxMapPosition: maxMapPosition ?? this.maxMapPosition,
      limit: limit ?? this.limit,
    );
  }

  /// Default filter for mobile app
  static ClanWarStatsFilter defaultFilter() {
    return ClanWarStatsFilter(
      startDate: DateTime.now().subtract(const Duration(days: 180)), // 6 months ago
      endDate: DateTime.now(),
      warType: "all",
      limit: 50,
    );
  }

  /// Check if any filters are active (not default)
  bool hasActiveFilters() {
    return ownTownHall != null ||
        enemyTownHall != null ||
        (ownTownHalls != null && ownTownHalls!.isNotEmpty) ||
        (enemyTownHalls != null && enemyTownHalls!.isNotEmpty) ||
        sameTownHall ||
        warType != "all" ||
        (warTypes != null && warTypes!.isNotEmpty) ||
        freshAttacksOnly != null ||
        minStars != null ||
        maxStars != null ||
        (allowedStars != null && allowedStars!.isNotEmpty) ||
        minDestruction != null ||
        maxDestruction != null ||
        minMapPosition != null ||
        maxMapPosition != null;
  }

  /// Get filter summary text
  String getFilterSummary() {
    List<String> filters = [];
    
    if (ownTownHalls != null && ownTownHalls!.isNotEmpty) {
      if (ownTownHalls!.length == 1) {
        filters.add("TH${ownTownHalls!.first} attacks");
      } else {
        filters.add("TH${ownTownHalls!.join(', ')} attacks");
      }
    } else if (ownTownHall != null) {
      filters.add("TH$ownTownHall attacks");
    }
    
    if (enemyTownHalls != null && enemyTownHalls!.isNotEmpty) {
      if (enemyTownHalls!.length == 1) {
        filters.add("vs TH${enemyTownHalls!.first}");
      } else {
        filters.add("vs TH${enemyTownHalls!.join(', ')}");
      }
    } else if (enemyTownHall != null) {
      filters.add("vs TH$enemyTownHall");
    }
    
    if (sameTownHall) filters.add("Same TH only");
    
    if (warTypes != null && warTypes!.isNotEmpty) {
      if (warTypes!.length == 1) {
        filters.add("${warTypes!.first.toUpperCase()} wars");
      } else {
        filters.add("${warTypes!.map((w) => w.toUpperCase()).join(', ')} wars");
      }
    } else if (warType != "all") {
      filters.add("${warType.toUpperCase()} wars");
    }
    
    if (freshAttacksOnly == true) filters.add("Fresh attacks only");
    
    if (allowedStars != null && allowedStars!.isNotEmpty) {
      if (allowedStars!.length == 1) {
        filters.add("${allowedStars!.first} ⭐ only");
      } else {
        filters.add("${allowedStars!.join(', ')} ⭐ only");
      }
    } else if (minStars != null || maxStars != null) {
      if (minStars != null && maxStars != null) {
        filters.add("$minStars-$maxStars stars");
      } else if (minStars != null) {
        filters.add("$minStars+ stars");
      } else {
        filters.add("≤$maxStars stars");
      }
    }
    
    if (minDestruction != null || maxDestruction != null) {
      if (minDestruction != null && maxDestruction != null) {
        filters.add("${minDestruction!.toStringAsFixed(0)}-${maxDestruction!.toStringAsFixed(0)}% destruction");
      } else if (minDestruction != null) {
        filters.add("${minDestruction!.toStringAsFixed(0)}%+ destruction");
      } else {
        filters.add("≤${maxDestruction!.toStringAsFixed(0)}% destruction");
      }
    }
    
    return filters.isEmpty ? "No filters applied" : filters.join(", ");
  }
}

File: expo/src/features/clan/models/clan-war-stats-filter.ts
export interface ClanWarStatsFilterOptions {
  readonly startDate?: Date | null;
  readonly endDate?: Date | null;
  readonly ownTownHall?: number | null;
  readonly enemyTownHall?: number | null;
  readonly ownTownHalls?: readonly number[] | null;
  readonly enemyTownHalls?: readonly number[] | null;
  readonly sameTownHall?: boolean;
  readonly warType?: string;
  readonly warTypes?: readonly string[] | null;
  readonly freshAttacksOnly?: boolean | null;
  readonly minStars?: number | null;
  readonly maxStars?: number | null;
  readonly allowedStars?: readonly number[] | null;
  readonly minDestruction?: number | null;
  readonly maxDestruction?: number | null;
  readonly minMapPosition?: number | null;
  readonly maxMapPosition?: number | null;
  readonly limit?: number;
}

export class ClanWarStatsFilter {
  readonly startDate: Date | null;
  readonly endDate: Date | null;
  readonly ownTownHall: number | null;
  readonly enemyTownHall: number | null;
  readonly ownTownHalls: readonly number[] | null;
  readonly enemyTownHalls: readonly number[] | null;
  readonly sameTownHall: boolean;
  readonly warType: string;
  readonly warTypes: readonly string[] | null;
  readonly freshAttacksOnly: boolean | null;
  readonly minStars: number | null;
  readonly maxStars: number | null;
  readonly allowedStars: readonly number[] | null;
  readonly minDestruction: number | null;
  readonly maxDestruction: number | null;
  readonly minMapPosition: number | null;
  readonly maxMapPosition: number | null;
  readonly limit: number;

  constructor(options: ClanWarStatsFilterOptions = {}) {
    this.startDate = options.startDate ?? null;
    this.endDate = options.endDate ?? null;
    this.ownTownHall = options.ownTownHall ?? null;
    this.enemyTownHall = options.enemyTownHall ?? null;
    this.ownTownHalls = options.ownTownHalls ?? null;
    this.enemyTownHalls = options.enemyTownHalls ?? null;
    this.sameTownHall = options.sameTownHall ?? false;
    this.warType = options.warType ?? 'all';
    this.warTypes = options.warTypes ?? null;
    this.freshAttacksOnly = options.freshAttacksOnly ?? null;
    this.minStars = options.minStars ?? null;
    this.maxStars = options.maxStars ?? null;
    this.allowedStars = options.allowedStars ?? null;
    this.minDestruction = options.minDestruction ?? null;
    this.maxDestruction = options.maxDestruction ?? null;
    this.minMapPosition = options.minMapPosition ?? null;
    this.maxMapPosition = options.maxMapPosition ?? null;
    this.limit = options.limit ?? 50;
  }

  toJson(): Record<string, unknown> {
    const data: Record<string, unknown> = { limit: this.limit, same_th: this.sameTownHall };
    if (this.warTypes?.length && !this.warTypes.includes('all')) data.type = this.warTypes;
    else if (this.warType !== 'all') data.type = this.warType;
    if (this.startDate) data.timestamp_start = Math.floor(this.startDate.getTime() / 1000);
    if (this.endDate) data.timestamp_end = Math.floor(this.endDate.getTime() / 1000);
    if (this.ownTownHalls?.length) data.own_th = this.ownTownHalls;
    else if (this.ownTownHall !== null) data.own_th = this.ownTownHall;
    if (this.enemyTownHalls?.length) data.enemy_th = this.enemyTownHalls;
    else if (this.enemyTownHall !== null) data.enemy_th = this.enemyTownHall;
    if (this.freshAttacksOnly !== null) data.fresh_only = this.freshAttacksOnly;
    if (this.allowedStars?.length) data.stars = this.allowedStars;
    else {
      if (this.minStars !== null) data.min_stars = this.minStars;
      if (this.maxStars !== null) data.max_stars = this.maxStars;
    }
    if (this.minDestruction !== null) data.min_destruction = this.minDestruction;
    if (this.maxDestruction !== null) data.max_destruction = this.maxDestruction;
    if (this.minMapPosition !== null) data.map_position_min = this.minMapPosition;
    if (this.maxMapPosition !== null) data.map_position_max = this.maxMapPosition;
    return data;
  }

  copyWith(options: ClanWarStatsFilterOptions): ClanWarStatsFilter {
    return new ClanWarStatsFilter({
      startDate: hasOption(options, 'startDate') ? options.startDate : this.startDate,
      endDate: hasOption(options, 'endDate') ? options.endDate : this.endDate,
      ownTownHall: hasOption(options, 'ownTownHall') ? options.ownTownHall : this.ownTownHall,
      enemyTownHall: hasOption(options, 'enemyTownHall')
        ? options.enemyTownHall
        : this.enemyTownHall,
      ownTownHalls: hasOption(options, 'ownTownHalls') ? options.ownTownHalls : this.ownTownHalls,
      enemyTownHalls: hasOption(options, 'enemyTownHalls')
        ? options.enemyTownHalls
        : this.enemyTownHalls,
      sameTownHall: options.sameTownHall ?? this.sameTownHall,
      warType: options.warType ?? this.warType,
      warTypes: hasOption(options, 'warTypes') ? options.warTypes : this.warTypes,
      freshAttacksOnly: hasOption(options, 'freshAttacksOnly')
        ? options.freshAttacksOnly
        : this.freshAttacksOnly,
      minStars: hasOption(options, 'minStars') ? options.minStars : this.minStars,
      maxStars: hasOption(options, 'maxStars') ? options.maxStars : this.maxStars,
      allowedStars: hasOption(options, 'allowedStars') ? options.allowedStars : this.allowedStars,
      minDestruction: hasOption(options, 'minDestruction')
        ? options.minDestruction
        : this.minDestruction,
      maxDestruction: hasOption(options, 'maxDestruction')
        ? options.maxDestruction
        : this.maxDestruction,
      minMapPosition: hasOption(options, 'minMapPosition')
        ? options.minMapPosition
        : this.minMapPosition,
      maxMapPosition: hasOption(options, 'maxMapPosition')
        ? options.maxMapPosition
        : this.maxMapPosition,
      limit: options.limit ?? this.limit,
    });
  }

  static defaultFilter(now = new Date()): ClanWarStatsFilter {
    return new ClanWarStatsFilter({
      startDate: new Date(now.getTime() - 180 * 24 * 60 * 60 * 1000),
      endDate: now,
    });
  }

  hasActiveFilters(): boolean {
    return Boolean(
      this.ownTownHall !== null ||
      this.enemyTownHall !== null ||
      this.ownTownHalls?.length ||
      this.enemyTownHalls?.length ||
      this.sameTownHall ||
      this.warType !== 'all' ||
      this.warTypes?.length ||
      this.freshAttacksOnly !== null ||
      this.minStars !== null ||
      this.maxStars !== null ||
      this.allowedStars?.length ||
      this.minDestruction !== null ||
      this.maxDestruction !== null ||
      this.minMapPosition !== null ||
      this.maxMapPosition !== null,
    );
  }

  getFilterSummary(): string {
    const filters: string[] = [];
    if (this.ownTownHalls?.length) filters.push(`TH${this.ownTownHalls.join(', ')} attacks`);
    else if (this.ownTownHall !== null) filters.push(`TH${this.ownTownHall} attacks`);
    if (this.enemyTownHalls?.length) filters.push(`vs TH${this.enemyTownHalls.join(', ')}`);
    else if (this.enemyTownHall !== null) filters.push(`vs TH${this.enemyTownHall}`);
    if (this.sameTownHall) filters.push('Same TH only');
    if (this.warTypes?.length)
      filters.push(`${this.warTypes.map((value) => value.toUpperCase()).join(', ')} wars`);
    else if (this.warType !== 'all') filters.push(`${this.warType.toUpperCase()} wars`);
    if (this.freshAttacksOnly === true) filters.push('Fresh attacks only');
    if (this.allowedStars?.length) filters.push(`${this.allowedStars.join(', ')} ⭐ only`);
    else if (this.minStars !== null && this.maxStars !== null)
      filters.push(`${this.minStars}-${this.maxStars} stars`);
    else if (this.minStars !== null) filters.push(`${this.minStars}+ stars`);
    else if (this.maxStars !== null) filters.push(`≤${this.maxStars} stars`);
    if (this.minDestruction !== null && this.maxDestruction !== null)
      filters.push(
        `${this.minDestruction.toFixed(0)}-${this.maxDestruction.toFixed(0)}% destruction`,
      );
    else if (this.minDestruction !== null)
      filters.push(`${this.minDestruction.toFixed(0)}%+ destruction`);
    else if (this.maxDestruction !== null)
      filters.push(`≤${this.maxDestruction.toFixed(0)}% destruction`);
    return filters.length ? filters.join(', ') : 'No filters applied';
  }
}

function hasOption(
  options: ClanWarStatsFilterOptions,
  key: keyof ClanWarStatsFilterOptions,
): boolean {
  return Object.prototype.hasOwnProperty.call(options, key);
}
